File: Cloud Sheets Samples/Google Sheets/python/googlesheets.py
# 
# Cloud Sheets 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of Cloud Sheets in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/cloudsheets
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from cloudsheets import *

input = sys.hexversion < 0x03000000 and raw_input or input


def ensureArg(args, prompt, index):
    if len(args) <= index:
        while len(args) <= index:
            args.append(None)
        args[index] = input(prompt)
    elif args[index] is None:
        args[index] = input(prompt)


def get_arg_value(flag):
    if flag in sys.argv:
        index = sys.argv.index(flag) + 1
        if index < len(sys.argv):
            return sys.argv[index]
    return None 

OAUTH_SCOPES = "https://www.googleapis.com/auth/spreadsheets https://www.googleapis.com/auth/drive"
OAUTH_AUTH_URL = "https://accounts.google.com/o/oauth2/auth"
OAUTH_TOKEN_UTL = "https://accounts.google.com/o/oauth2/token"
OAUTH_CLIENT_ID = get_arg_value("/c")
OAUTH_SECRET_ID = get_arg_value("/s")

gsheets = GoogleSheets()
changes_made = False

if len(sys.argv) < 4:
    print("Usage: python googlesheets.py /c client /s secret\n" +
            " /c The Id of the client assigned when registering the application.\n" +
            " /s The secret value for the client assigned when registering the application.")


def authenticate(client, secret):
    oauth = OAuth()
    oauth.client_profile = 0
    oauth.grant_type = 0
    oauth.client_id = client
    oauth.client_secret = secret
    oauth.authorization_scope = OAUTH_SCOPES
    oauth.server_auth_url = OAUTH_AUTH_URL
    oauth.server_token_url = OAUTH_TOKEN_UTL

    try:
        gsheets.authorization = oauth.get_authorization()
        print("\nAuthorized")
    except CloudSpreadsheetsError as ex:
        print(ex)
        
def display_menu():
    print("\r\n?\t-\tHelp\n" +
          "ls\t-\tList Spreadsheets\n" +
          "lssh\t-\tList Sheets\n" +
          "cs\t-\tCreate Spreadsheet\n" +
          "ash\t-\tAdd Sheet\n" +
          "us\t-\tUpdate Spreadsheet\n" +
          "ush\t-\tUpdate Sheet\n" +
          "ds\t-\tDelete Spreadsheet\n" +
          "dsh\t-\tDelete Sheet\n" +
          "exp\t-\tExport Spreadsheet\n" +
          "q\t-\tQuit\n")
    
def spreadsheet_on_spreadsheet_list(e):
    print("ID: " + e.id)
    print("Name: " + e.name)
    
try: 
    gsheets = GoogleSheets() 
    gsheets.on_spreadsheet_list = spreadsheet_on_spreadsheet_list
    authenticate(OAUTH_CLIENT_ID, OAUTH_SECRET_ID)
    
    print("\nListing Spreadsheets...")
    gsheets.list_spreadsheets()
    
    display_menu()
    
    while True:
        buffer = input("Enter command: ")
        
        try:
            match buffer:
                case "ls":
                    gsheets.list_spreadsheets()
                    
                case "lssh":
                    current_spreadsheet_id = input("Enter the Spreadsheet Id: ")
                    gsheets.load_spreadsheet(current_spreadsheet_id)
                    
                    for i in range((gsheets.sheets_count)):
                        print("Id: " + str(gsheets.get_sheet_sheet_id(i)))
                        print("Title: " + gsheets.get_sheet_title(i))
                        print("Index: " + str(gsheets.get_sheet_sheet_index(i)))
                        print("Total Columns: " + str(gsheets.get_sheet_total_column_count(i)))
                        print("Total Rows: " + str(gsheets.get_sheet_total_row_count(i)))

                case "cs":
                    gsheets.create_spreadsheet(input("Enter Spreadsheet Name: "), input("Enter Locale (optional): "), input("Enter Timezone (optional): "))
                    print("Success...")
                    
                case "ash":
                    current_spreadsheet_id = input("Enter the Spreadsheet Id: ")
                    gsheets.load_spreadsheet(current_spreadsheet_id)
                    gsheets.add_sheet(int(input("Enter the index where sheet should be added: ")), input("Enter the title: "))
                
                case "us":
                    changes_made = False
                    current_spreadsheet_id = input("Enter the Spreadsheet Id: ")
                    gsheets.load_spreadsheet(current_spreadsheet_id)
                    
                    if (input("Current name: " + gsheets.get_spreadsheet_name(0) + "\nChange? y/n: ") or "n") == "y":
                        gsheets.set_spreadsheet_name(0, input("Enter a new name: "))
                        changes_made = True
                    
                    if (input("Current locale: " + gsheets.get_spreadsheet_locale(0) + "\nChange? y/n: ") or "n") == "y":
                        gsheets.set_spreadsheet_name(0, input("Enter a new locale (e.g. en_US): "))
                        changes_made = True
                        
                    if (input("Current timezone: " + gsheets.get_spreadsheet_locale(0) + "\nChange? y/n: ") or "n") == "y":
                        gsheets.set_spreadsheet_name(0, input("Enter a new locale (e.g. America/New_York): "))
                        changes_made = True  
                        
                    if changes_made:
                            gsheets.update_spreadsheet()
                            print("Success...")
                            changes_made = False 
                    else:
                        print("Couldn't update Spreadsheet.")
                        
                case "ush":
                    current_spreadsheet_id = input("Enter Spreadsheet Id: ")
                    current_sheet_id = int(input(f"Enter Sheet Id (Should be in Spreadsheet with Id '{current_spreadsheet_id}'): "))
                    
                    gsheets.load_spreadsheet(current_spreadsheet_id)
                    gsheets.load_sheet(current_sheet_id)
                    loaded_sheet_title = gsheets.get_sheet_title(0)
                    loaded_sheet_index = gsheets.get_sheet_sheet_index(0)
                    
                    new_title = input("Current title: " + loaded_sheet_title + "\nEnter new title (or press Enter to keep the same): ") or loaded_sheet_title
                    new_index = int(input("Current index: " + str(loaded_sheet_index) + "\nEnter new index: ") or loaded_sheet_index)
                
                    if new_title != loaded_sheet_title or new_index != loaded_sheet_index:
                        gsheets.set_sheet_title(0, new_title)
                        gsheets.set_sheet_sheet_index(0, new_index)
                        changes_made = True
                    
                    if (input("Total rows: " + str(gsheets.get_sheet_total_row_count(0)) + "\nChange? y/n: ") or "n") == "y":
                        gsheets.set_sheet_total_row_count(0, int(input("Enter new total row count: ")))
                        changes_made = True
                        
                    if(input("Total columns: " + str(gsheets.get_sheet_total_column_count(0)) + "\nChange? y/n: ") or "n") == "y":
                        gsheets.set_sheet_total_column_count(0, int(input("Enter new total columns count: ")))
                        changes_made = True
                        
                    if changes_made:
                            gsheets.update_sheet()
                            print("Success...")
                            changes_made = False 
                    else:
                        print("Couldn't update Spreadsheet.")
                                    
                case "ds":
                    gsheets.delete_spreadsheet(input("Enter Spreadsheet Id: "))
                    print("Success...")
                
                case "dsh":
                    current_spreadsheet_id = input("Enter Spreadsheet Id: ")
                    current_sheet_id = int(input(f"Enter Sheet Id (Should be in Spreadsheet with Id '{current_spreadsheet_id}'): "))
                    gsheets.delete_sheet(current_sheet_id)
                
                case "exp":
                    print("Supported file extensions: csv, ods, pdf, tsv, xlsx, zip")
                    current_spreadsheet_id = input("\nEnter Spreadsheet Id: ")
                    file_name = input("Enter full filename (e.g. myFile.zip): ")
                    file_path = input("Enter save directory: ") + f"\\{file_name}"
                    
                    print(file_path)
                    gsheets.export(current_spreadsheet_id, file_path)
                    print("Success...")
                                    
                case "?":
                    display_menu()
                
                case "q":
                    break
                
                case _:
                    print("Not a known command!")
                
        except CloudSpreadsheetsError as ex:
            print(ex)
            
except CloudSpreadsheetsError as ex:
    print(ex)

File: Cloud Sheets Samples/Google Sheets/java/googlesheets.java
/*
 * Cloud Sheets 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Sheets in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudsheets
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import cloudsheets.*;


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Map;

public class googlesheets extends ConsoleDemo {
    private static final String oAuthScope = "https://www.googleapis.com/auth/spreadsheets https://www.googleapis.com/auth/drive";
    private static final String oAuthAuthUrl = "https://accounts.google.com/o/oauth2/auth";
    private static final String oAuthTokenUrl = "https://accounts.google.com/o/oauth2/token";
    private static GoogleSheets gsheets;
    public static String currentSpreadsheetId = "";
    public static int currentSheetId = -1;
    public boolean changesMade = false;
    public static GSSheet currentSheet = null;
    private static Map<String, String> myArgs;


    public static void main(String[] args) {
        if(args.length <4){
            System.out.println("usage: GoogleSheetsDemo.java -c client -s secret\n" +
                    " -c The Id of the client assigned when registering the application.\n" +
                    " -s The secret value for the client assigned when registering the application.");
            return;
        }

        try{
            String buffer;
            myArgs = parseArgs(args);
            gsheets = new GoogleSheets();
            gsheets.addGoogleSheetsEventListener(new Events());
            authenticate(myArgs.get("c"), myArgs.get("s"));

            System.out.println("Listing Spreadsheets...");
            gsheets.listSpreadsheets();
            displayMenu();

            while (true){
                buffer = prompt("\nEnter command");

                try{
                    switch (buffer){
                        case "ls":
                            gsheets.listSpreadsheets();
                            break;
                        case "lssh":
                            currentSpreadsheetId = prompt("Enter the Spreadsheet Id");
                            gsheets.loadSpreadsheet(currentSpreadsheetId);
                            for(GSSheet sheet : gsheets.getSheets()){
                                System.out.println("Id: " + sheet.getSheetId());
                                System.out.println("Title: " + sheet.getTitle());
                                System.out.println("Index: " + sheet.getSheetIndex());
                                System.out.println("Total Columns: " + sheet.getTotalColumnCount());
                                System.out.println("Total Rows: " + sheet.getTotalRowCount());
                            }
                            break;
                        case "cs":
                            gsheets.createSpreadsheet(
                                    prompt("Enter Spreadsheet name"),
                                    prompt("Enter locale (optional)"),
                                    prompt("Enter timezone (optional)"));
                            System.out.println("Success...");
                            break;
                        case "ash":
                            currentSpreadsheetId = prompt("Enter Spreadsheet Id");
                            gsheets.loadSpreadsheet(currentSpreadsheetId);
                            gsheets.addSheet(
                                    Integer.parseInt(prompt("Enter the index where sheet should be added")),
                                    prompt("Enter sheet title"));
                            System.out.println("Success...");
                            break;
                        case "us":
                            boolean changesMade = false;
                            currentSpreadsheetId = prompt("Enter Spreadsheet Id");
                            gsheets.loadSpreadsheet(currentSpreadsheetId);
                            GSSpreadsheet spreadsheet = gsheets.getSpreadsheets().get(0);
                            if (prompt("Current name: " + spreadsheet.getName() + "\nChange? y/n: ", "n")
                                    .equalsIgnoreCase("y")) {
                                spreadsheet.setName(prompt("Enter new name"));
                                changesMade = true;
                            }
                            if (prompt("Current locale: " + spreadsheet.getLocale() + "\nChange? y/n", "n")
                                    .equalsIgnoreCase("y")) {
                                spreadsheet.setLocale(prompt("Enter new locale (e.g. en_US)"));
                                changesMade = true;
                            }
                            if (prompt("Current timezone:" + spreadsheet.getTimeZone() + "\nChange? y/n", "n")
                                    .equalsIgnoreCase("y")) {
                                spreadsheet.setTimeZone(prompt("Enter new timezone (e.g. America/New_York)"));
                                changesMade = true;
                            }
                            if (changesMade) {
                                gsheets.updateSpreadsheet();
                                System.out.println("Success...");
                            } else {
                                System.out.println("Couldn't update Spreadsheet");
                            }
                            break;
                        case "ush":
                            currentSpreadsheetId = prompt("Enter Spreadsheet Id");
                            currentSheetId = Integer.parseInt(prompt("Enter Sheet Id (Should be in Spreadsheet with Id: " + currentSpreadsheetId + ")"));
                            gsheets.loadSpreadsheet(currentSpreadsheetId);
                            gsheets.loadSheet(currentSheetId);
                            GSSheet loadedSheet = gsheets.getSheets().get(0);
                            String newTitle = prompt("Current title: " + loadedSheet.getTitle() +
                                    " Enter new title (or press Enter to keep the same): ", loadedSheet.getTitle());
                            int newIndex = Integer.parseInt(prompt("Current index: " + loadedSheet.getSheetIndex() +
                                    " Enter new index: ", String.valueOf(loadedSheet.getSheetIndex())));
                            if (!newTitle.equals(loadedSheet.getTitle()) || newIndex != loadedSheet.getSheetIndex()) {
                                loadedSheet.setTitle(newTitle);
                                loadedSheet.setSheetIndex(newIndex);
                            }
                            boolean sheetChangesMade = false;
                            if (prompt("Total rows: " + loadedSheet.getTotalRowCount() + "\nChange? y/n: ", "n")
                                    .equalsIgnoreCase("y")) {
                                loadedSheet.setTotalRowCount(Integer.parseInt(prompt("Enter new total row count")));
                                sheetChangesMade = true;
                            }
                            if (prompt("Total columns: " + loadedSheet.getTotalColumnCount() + "\nChange? y/n", "n")
                                    .equalsIgnoreCase("y")) {
                                loadedSheet.setTotalColumnCount(Integer.parseInt(prompt("Enter new total columns count")));
                                sheetChangesMade = true;
                            }
                            if (sheetChangesMade) {
                                gsheets.updateSheet();
                                System.out.println("Success...");
                            } else {
                                System.out.println("Couldn't update Sheet");
                            }
                            break;
                        case "ds":
                            gsheets.deleteSpreadsheet(prompt("Enter Spreadsheet Id"));
                            System.out.println("Success...");
                            break;
                        case "dsh":
                            currentSpreadsheetId = prompt("Enter Spreadsheet Id");
                            currentSheetId = Integer.parseInt(prompt("Enter Sheet Id (Should be in Spreadsheet with Id: " + currentSpreadsheetId + ")"));
                            gsheets.deleteSheet(currentSheetId);
                            break;
                        case "exp":
                            System.out.println("Supported file extensions: csv, ods, pdf, tsv, xlsx, zip");
                            currentSpreadsheetId = prompt("Enter Spreadsheet Id");
                            String fileName = prompt("Enter full filename (e.g. myFile.zip)");
                            String filePath = prompt("Enter save directory") + File.separator + fileName;
                            System.out.println(filePath);
                            gsheets.export(currentSpreadsheetId, filePath);
                            System.out.println("Success...");
                            break;
                        case "q":
                            System.exit(0);
                            break;
                        default:
                            System.out.println("Not a known command!");
                            break;
                    }

                }
                catch(CloudSpreadsheetsException ex){
                    System.out.println(ex.getMessage());
                }
            }

        }
        catch(Exception ex){
            System.out.println(ex.getMessage());
        }
    }

    private static void authenticate(String client, String secret) throws Exception{
        OAuth oauth = new OAuth();

        if(client != null && secret != null){
            oauth.setClientId(client);
            oauth.setClientSecret(secret);
            oauth.setAuthorizationScope(oAuthScope);
            oauth.setServerAuthURL(oAuthAuthUrl);
            oauth.setServerTokenURL(oAuthTokenUrl);

            try{
                gsheets.setAuthorization(oauth.getAuthorization());
            }
            catch(CloudSpreadsheetsException ex){
                System.out.println(ex.getMessage());
            }
        }
        else {
            System.out.println("Client Id and Secret cannot be null!");
        }
    }

    private static void displayMenu()
    {
        System.out.println("\r\n?\t-\tHelp\n" +
                "ls\t-\tList Spreadsheets\n" +
                "lssh\t-\tList Sheets\n" +
                "cs\t-\tCreate Spreadsheet\n" +
                "ash\t-\tAdd Sheet\n" +
                "us\t-\tUpdate Spreadsheet\n" +
                "ush\t-\tUpdate Sheet\n" +
                "ds\t-\tDelete Spreadsheet\n" +
                "dsh\t-\tDelete Sheet\n" +
                "exp\t-\tExport Spreadsheet\n" +
                "q\t-\tQuit\n");
    }
}


class Events implements GoogleSheetsEventListener
{

    @Override
    public void error(GoogleSheetsErrorEvent googleSheetsErrorEvent) {

    }

    @Override
    public void log(GoogleSheetsLogEvent googleSheetsLogEvent) {

    }

    @Override
    public void spreadsheetList(GoogleSheetsSpreadsheetListEvent googleSheetsSpreadsheetListEvent) {
        System.out.println("ID: " + googleSheetsSpreadsheetListEvent.id);
        System.out.println("Name: " + googleSheetsSpreadsheetListEvent.name);
    }

    @Override
    public void SSLServerAuthentication(GoogleSheetsSSLServerAuthenticationEvent googleSheetsSSLServerAuthenticationEvent) {

    }

    @Override
    public void SSLStatus(GoogleSheetsSSLStatusEvent googleSheetsSSLStatusEvent) {

    }

    @Override
    public void transfer(GoogleSheetsTransferEvent googleSheetsTransferEvent) {

    }
}
class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "]" + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof CloudSheetsException) {
      System.out.print(" (" + ((CloudSheetsException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }

  /**
   * Takes a list of switch arguments or name-value arguments and turns it into a map.
   */
  static java.util.Map<String, String> parseArgs(String[] args) {
    java.util.Map<String, String> map = new java.util.HashMap<String, String>();
    
    for (int i = 0; i < args.length; i++) {
      // Add a key to the map for each argument.
      if (args[i].startsWith("-")) {
        // If the next argument does NOT start with a "-" then it is a value.
        if (i + 1 < args.length && !args[i + 1].startsWith("-")) {
          // Save the value and skip the next entry in the list of arguments.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), args[i + 1]);
          i++;
        } else {
          // If the next argument starts with a "-", then we assume the current one is a switch.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), "");
        }
      } else {
        // If the argument does not start with a "-", store the argument based on the index.
        map.put(Integer.toString(i), args[i].toLowerCase());
      }
    }
    return map;
  }
}



